File: cpp/grade-school/grade_school.cpp
#include "grade_school.h"

namespace grade_school {

bool school::add(std::string student_name, int grade) {
  if (is_already_registred(student_name)) {
    return false;
  }
  this->registred_students.insert(student_name);
  this->students_by_grade[grade].push_back(student_name);
  return true;
}

std::vector<std::string> school::grade(int value) const {
  if (!is_grade_initialized(value)) {
    return {};
  }
  std::vector<std::string> in_grade = this->students_by_grade.at(value);
  std::sort(in_grade.begin(), in_grade.end());
  return in_grade;
}

std::map<int, std::vector<std::string>> school::roster() const {
  std::map<int, std::vector<std::string>> roster_;
  for (const auto &pair : this->students_by_grade) {
    int key = pair.first;
    roster_[key] = this->grade(key);
  }
  return roster_;
}

bool school::is_already_registred(std::string name) const {
  return (this->registred_students.find(name) !=
          this->registred_students.end());
}

bool school::is_grade_initialized(int grade) const {
  return this->students_by_grade.find(grade) != this->students_by_grade.end();
}

} // namespace grade_school

File: lua/grade-school/grade-school.lua
--- A class representing a school, used to manage a list of students and their grades.
--- @class School
--- @field _students table<string, integer> A private table mapping student names to their corresponding grades.
local School <const> = {}

--- Creates a new `School` instance.
--- @return School -- The newly created `School` object.
function School:new()
    return setmetatable({ _students = {} }, School)
end

--- Adds a student with their name and grade to the school's roster.
--- @param self School
--- @param name string The name of the student.
--- @param grade integer The grade of the student.
--- @return boolean -- `true` if the student was added successfully, `false` if a student with
--- the same name already exists.
function School.add(self, name, grade)
    if self._students[name] ~= nil then
        return false
    end
    self._students[name] = grade
    return true
end

--- Returns a sorted list of all students on the school's roster.
--- The list is sorted primarily by grade in ascending order. Students in the same grade are
--- sorted alphabetically by name.
--- @param self School
--- @return string[] -- A table containing the names of all students.
function School.roster(self)
    local result <const> = {}

    for name, _ in pairs(self._students) do
        table.insert(result, name)
    end

    table.sort(result, function(a, b)
        if self._students[a] < self._students[b] then
            return true
        elseif self._students[a] == self._students[b] then
            return a < b
        else
            return false
        end
    end)

    return result
end

--- Returns a sorted list of all students in a specific grade.
--- The list is sorted alphabetically by student name.
--- @param self School
--- @param target integer The target grade to retrieve students from.
--- @return string[] -- A table containing the names of the students in the specified grade.
function School.grade(self, target)
    local result <const> = {}

    for name, grade in pairs(self._students) do
        if grade == target then
            table.insert(result, name)
        end
    end

    table.sort(result)

    return result
end

--- The `__index` metamethod ensures that method calls on `School` instances are correctly
--- redirected to the `School` class table.
School.__index = School

return School
